From ef1d91fb132e6d63d5ec3eceaff6fc15f7beb8a6 Mon Sep 17 00:00:00 2001
From: Paul Cornett <paulcor@users.noreply.github.com>
Date: Sun, 5 Jan 2025 14:01:04 -0800
Subject: [PATCH] Fix border size for wxStaticBox with GTK3

The GtkFrame border was not being accounted for in GetBordersForSizer().

See #25045.

(cherry picked from commit 12bb1123eb620c6541593465e3a628b2a4b2faa8)
---
 src/gtk/statbox.cpp | 18 ++++++++++++++++++
 1 file changed, 18 insertions(+)

diff --git a/src/gtk/statbox.cpp b/src/gtk/statbox.cpp
index f5a69afa5f94..50f893eccdb1 100644
--- a/src/gtk/statbox.cpp
+++ b/src/gtk/statbox.cpp
@@ -15,6 +15,8 @@
 
 #include "wx/gtk/private/wrapgtk.h"
 #include "wx/gtk/private/win_gtk.h"
+#include "wx/gtk/private/stylecontext.h"
+#include "wx/gtk/private/gtk3-compat.h"
 
 // constants taken from GTK sources
 #define LABEL_PAD 1
@@ -204,6 +206,22 @@ void wxStaticBox::GetBordersForSizer(int *borderTop, int *borderOther) const
         gtk_widget_get_preferred_width(label, NULL, &nat_width);
         gtk_widget_get_preferred_height_for_width(label, nat_width, borderTop, NULL);
     }
+    wxGtkStyleContext sc(GetContentScaleFactor());
+    sc.Add(GTK_TYPE_FRAME, "frame", "frame", NULL);
+    if (wx_is_at_least_gtk3(20))
+        sc.Add("border");
+    else
+    {
+        *borderOther = gtk_container_get_border_width(GTK_CONTAINER(m_widget));
+        *borderTop += *borderOther;
+    }
+    GtkBorder border;
+    gtk_style_context_get_border(sc, GTK_STATE_FLAG_NORMAL, &border);
+    *borderOther += border.left;
+    *borderTop += border.top;
+    gtk_style_context_get_padding(sc, GTK_STATE_FLAG_NORMAL, &border);
+    *borderOther += border.left;
+    *borderTop += border.top;
 #else
     gtk_widget_ensure_style(m_widget);
     const int border_width = GTK_CONTAINER(m_widget)->border_width;
